File: src/test/java/org/chase/telegram/filterchain/handlers/MessageFilterHandlerTest.java
package org.chase.telegram.filterchain.handlers;

import org.chase.telegram.filterchain.Handler;
import org.chase.telegram.filterchain.UpdateContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.telegram.telegrambots.meta.api.objects.Update;

import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class MessageFilterHandlerTest {

    @Mock
    Update updateMock;

    @Mock
    Handler handlerMock;

    @Mock
    UpdateContext updateContextMock;

    MessageFilterHandler sut = new MessageFilterHandler();

    @BeforeEach
    void init() {
        lenient().when(updateContextMock.getUpdate()).thenReturn(updateMock);
        lenient().when(handlerMock.call(updateContextMock)).thenReturn(updateContextMock);
    }

    @Test
    void shouldCallNextHandlerIfUpdateHasAMessage() {
        sut.setNext(handlerMock);
        when(updateMock.hasMessage()).thenReturn(true);

        sut.call(updateContextMock);

        verify(handlerMock, times(1)).call(updateContextMock);

    }

    @Test
    void shouldNotCallNextHandlerIfNoHandlerWasSet() {
        when(updateMock.hasMessage()).thenReturn(true);

        sut.call(updateContextMock);

        verifyNoInteractions(handlerMock);
    }

    @Test
    void shouldNotCallNextHandlerIfUpdateHasNoMessage() {
        sut.setNext(handlerMock);

        when(updateMock.hasMessage()).thenReturn(false);

        sut.call(updateContextMock);

        verifyNoInteractions(handlerMock);
    }
}

File: src/test/kotlin/org/chase/telegram/filterchain/handlers/MessageFilterHandlerTest.kt
package org.chase.telegram.filterchain.handlers

import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.ShouldSpec
import io.mockk.*
import org.chase.telegram.filterchain.Handler
import org.chase.telegram.filterchain.UpdateContext
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.meta.bots.AbsSender

class MessageFilterHandlerTest : ShouldSpec({
    val updateMock: Update = mockk()
    val handlerMock: Handler = spyk(TestHandler())

    val senderMock: AbsSender = mockk()

    val context = UpdateContext(updateMock, senderMock)

    every { updateMock.hasMessage() } returns true

    val sut = MessageFilterHandler()

    context("MessageHandlerFilter") {
        should("call next handler if update has a message") {
            sut.next = handlerMock
            sut.call(context)

            verify(exactly = 1) { handlerMock.call(context) }
            confirmVerified()
        }

        should("not call next handler if no handler was set") {
            sut.call(context)

            verify(exactly = 0) { handlerMock.call(context) }
            confirmVerified()
        }

        should("not call next handler if update has no message") {
            sut.next = handlerMock

            every { updateMock.hasMessage() } returns false
            sut.call(context)

            verify(exactly = 0) { handlerMock.call(context) }
            confirmVerified()
        }
    }

    isolationMode = IsolationMode.InstancePerTest
})